### PR TITLE
External links

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
As requested https://github.com/nextcloud/android/issues/555

Done so far
- fetch and display via api
- storing in database
- configurable by setup.xml
- quota and sidebar links have "burger menu" and allow opening sidebar
- setting webview does not allow sidebar and therefore shows back button

Here are some screenshots

screenshot_1491461915
screenshot_1491461920
screenshot_1491461931
screenshot_1491462053
screenshot_1491462058

### CODE REVIEW

**Comment by MorrisJobke** [comment]
> how to handle the language? --> waiting for confirmation that api is handling this

Yes - we have a server side setting for "all languages" and a specific one. The specific entries are only returned by the API if they match with the user setting.

File: src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+                    ProviderMeta.ProviderTableMeta.EXTERNAL_LINKS_NAME));
+            String url = cursor.getString(cursor.getColumnIndex(
+                    ProviderMeta.ProviderTableMeta.EXTERNAL_LINKS_URL));
+
+            externalLink = new ExternalLink(id, iconUrl, language, type, name, url);
+        }
+        return externalLink;
+    }
+
+    private void downloadPNGIcon(Context context, String iconUrl, SimpleTarget imageView, int placeholder) {
```
Comment (mario) [inline src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java]: Don't think this belong in a provider. Rather a helper/utils somewhere? :)

File: src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+                .with(context)
+                .load(iconUrl)
+                .centerCrop()
+                .placeholder(placeholder)
+                .error(placeholder)
+                .crossFade()
+                .into(imageView);
+    }
+
+    private void downloadSVGIcon(Context context, String iconUrl, SimpleTarget imageView, int placeholder,
```
Comment (mario) [inline src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java]: Don't think this belong in a provider. Rather a helper/utils somewhere? :)

File: src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        mContentResolver = contentResolver;
+    }
+
+    /**
+     * Stores an external link in database.
+     *
+     * @param externalLink object to store
+     * @return external link id, -1 if the insert process fails.
+     */
+    public long storeExternalLink(ExternalLink externalLink) {
```
Comment (mario) [inline src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java]: Doesn't look like we store which cloudID the link belongs to? We should definitely do that so we know for which account links are for :D

File: src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+        Uri uri = Uri.parse(iconUrl);
+        requestBuilder
+                .diskCacheStrategy(DiskCacheStrategy.SOURCE)
+                .load(uri)
+                .into(imageView);
+    }
+
+
+    public void downloadIcon(Context context, String iconUrl, SimpleTarget imageView, int placeholder,
```
Comment (mario) [inline src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java]: See above :P

File: src/main/java/com/owncloud/android/ui/activity/ExternalSiteWebView.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                    }
+                } else {
+                    Intent settingsIntent = new Intent(getApplicationContext(), Preferences.class);
+                    startActivity(settingsIntent);
+                }
+            retval = true;
+            break;
+
+            default:
+                retval = super.onOptionsItemSelected(item);
```
Comment (mario) [inline src/main/java/com/owncloud/android/ui/activity/ExternalSiteWebView.java]: break? :)

File: src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java (hunk 1 above)
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java]: Done

File: src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java (hunk 2 above)
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java]: 👍

File: src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java (hunk 4 above)
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/datamodel/ExternalLinksProvider.java]: Refactored ;-)

File: src/main/java/com/owncloud/android/ui/activity/ExternalSiteWebView.java (hunk 5 above)
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/activity/ExternalSiteWebView.java]: As good as codacy ;-)

**Comment by tobiasKaminsky** [comment]
@mario please review again :-)

**Comment by mario** [comment]
👍

**Pull request merged**